### PR TITLE
interactive:  add (multi-level tree) traversal functionality

### PR DESCRIPTION
**Pull request by dshaaban01**
add (multi-level tree) traversal functionality

### CODE REVIEW

File: xdsl/interactive/app.py [hunk 1]
```diff
@@ -345,16 +379,20 @@ def add_pass_with_arguments_to_pass_pipeline(concatenated_arg_val: str) -> None:
                         f"{selected_pass_value.name}{{{concatenated_arg_val}}}"
                     )
                 )[0]
-                self.pass_pipeline = (
-                    *self.pass_pipeline,
+                # self.pass_pipeline = (
+                #     *self.pass_pipeline,
+                #     (selected_pass_value, new_pass_with_arguments),
+                # )
```
Comment (superlopuh) [inline xdsl/interactive/app.py:385]: commented code

File: xdsl/interactive/app.py [hunk 2]
```diff
@@ -135,6 +135,10 @@ class InputApp(App[None]):
     DataTable displaying the diff of operation names and counts of the input and output
     text areas.
     """
+    current_argument_pass: tuple[tuple[type[ModulePass], PipelinePassSpec], ...] = ()
+    """
+    Saves a tuple containing the constructed ModulePass, PipelinePassSpec tuple of a pass with arguments requiring user input.
+    """
```
Comment (superlopuh) [inline xdsl/interactive/app.py:141]: I thought we had agreed to remove this, I'm still not sure what breaks if this variable is removed. It feels scary to me, complexity-wise, to add this variable without a well-defined meaning. I don't know what a "current argument pass" is.

Comment (dshaaban01) [reply xdsl/interactive/app.py]: I removed ability to expand argument passes. What do you think now?

File: xdsl/interactive/app.py [hunk 3]
```diff
@@ -408,19 +411,63 @@ def update_pass_pipeline(
...
+            return
+
+        # get instance
+        selected_pass_value, selected_pass_spec = expanded_node.data
+
+        # if expanded_pass has arguments, do not allow node expansion
+        if fields(selected_pass_value) and selected_pass_spec is None:
+            return
+
+        else:
```
Comment (superlopuh) [inline xdsl/interactive/app.py:453]: no need for else here, can just dedent to make the structure a little clearer

**Review by superlopuh — APPROVED** [review]

**Pull request merged**